docs/guides/state_management.md

```diff
@@ -0,0 +1,118 @@
+
+```py
+@me.page()
+def page():
+    state = me.state(State)
+    me.text(state.val)
+```
+
+## How State Works
+
+`me.stateclass` is a class decorator which tells Mesop that this class can be retrieved using the `me.state` method, which will return the state instance for the current user session.
+
+> If you are familiar with the dependency injection pattern, Mesop's stateclass and state API is essentially a minimalist dependency injection system which scopes the state object to the lifetime of a user session.
+
+Under the hood, Mesop is sending the state back and forth between the server and browser client so everything in a state class, recursively, must be serializable.
+
+## Multiple state classes
+
+You can use multiple classes to store state for the current user session.
+
+Using different state classes for different pages or components can help make your app easier to maintain and more modular.
+
+```py
+@me.stateclass
+class PageAState:
+    ...
+
+@me.stateclass
+class PageBState:
+    ...
+
+@me.page(path="/a")
+def page_a():
+    state = me.state(PageAState)
+    ...
+
+@me.page(path="/b")
+def page_b():
+    state = me.state(PageBState)
+    ...
+```
+
+Under the hood, Mesop is managing state classes based on the identity (e.g. module name and class name) of the state class, which means that you could have two state classes named "State", but if they are in different modules, then they will be treated as separate state, which is what you would expect.
+
+## Nested State
+
+You can also have classes inside of a state class as long as everything is serializable:
+
+```python
+class NestedState:
+  val: str
+
+@me.stateclass
+class State:
+  nested: NestedState
+
+def app():
+  state = me.state(State)
+```
+
+> Note: you only need to decorate the top-level state class with `@me.stateclass`. All the nested state classes will automatically be wrapped.
+
+### Nested State and dataclass
+
+Sometimes, you may want to explicitly decorate the nested state class with `dataclass` because in the previous example, you couldn't directly instantiate `NestedState`.
+
+If you wanted to use NestedState as a general dataclass, you can do the following:
+
+```python
+@dataclass
+class NestedState:
+  val: str = ""
+
+@me.stateclass
+class State:
+  nested: NestedState
+
+def app():
+  state = me.state(State)
+```
+
+> Reminder: because dataclass do not have default values, you will need to explicitly set default values, otherwise Mesop will not be able to instantiate an empty version of the class.
+
+Now, if you have an event handler function, you can do the following:
+
+```py
+def on_click(e):
+    response = call_api()
+    state = me.state(State)
+    state.nested = NestedState(val=response.text)
+```
+
+If you didn't explicitly annotate NestedState as a dataclass, then you would get an error instantiating NestedState because there's no initializer defined.
+
+## Tips
+
+### Avoid mutable default values (e.g. list)
```

Comment:
[nit] I think if you tried to set a default like `var1: list[int] = [1]`, you'd get an error I think.

The dataclass solution is to use `field(default_factory=lambda: [1])` which would create a separate instance each time. Or also can use the onload.

So I feel this tip is could be worded a bit differently. Because we can set defaults values for mutable values.  Just not the more intuitive way,

Reply:
thx. added the solution to use field(default_factory). `onload` can produce a similar effect, but it's not quite the same since it can get executed multiple times for a user session (e.g. navigate to the same page twice).